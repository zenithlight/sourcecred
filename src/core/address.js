// @flow

import stringify from "json-stable-stringify";
import deepFreeze from "deep-freeze";

import * as MapUtil from "../util/map";
import * as C from "../util/combo";

export interface AddressModule<Address: string> {
  /**
   * Assert at runtime that the provided address is actually a valid
   * address of this kind, throwing an error if it is not. If `what` is
   * provided, it will be included in the error message.
   */
  assertValid(address: Address, what?: string): void;

  /**
   * Assert at runtime that the provided array is a valid array of
   * address parts (i.e., a valid input to `fromParts`), throwing an
   * error if it is not. If `what` is provided, it will be included in
   * the error message.
   */
  assertValidParts(parts: $ReadOnlyArray<string>, what?: string): void;

  /**
   * The empty address (the identity for `append`). Equivalent to
   * `fromParts([])`.
   */
  empty: Address;

  /**
   * Convert an array of address parts to an address. The input must be
   * a non-null array of non-null strings, none of which contains the
   * NUL character. This is the inverse of `toParts`.
   */
  fromParts(parts: $ReadOnlyArray<string>): Address;

  /**
   * Convert an address to the array of parts that it represents. This
   * is the inverse of `fromParts`.
   */
  toParts(address: Address): string[];

  /**
   * Pretty-print an address. The result will be human-readable and
   * contain only printable characters. Clients should not make any
   * assumptions about the format.
   */
  toString(address: Address): string;

  /**
   * Construct an address by extending the given address with the given
   * additional components. This function is equivalent to:
   *
   *     return fromParts([...toParts(address), ...components]);
   *
   * but may be more efficient.
   */
  append(address: Address, ...components: string[]): Address;

  /**
   * Test whether the given address has the given prefix. This function
   * is equivalent to:
   *
   *     const prefixParts = toParts(prefix);
   *     const addressParts = toParts(address);
   *     const actualPrefix = addressParts.slice(0, prefixParts.length);
   *     return deepEqual(prefix, actualPrefix);
   *
   * (where `deepEqual` checks value equality on arrays of strings), but
   * may be more efficient.
   *
   * Note that this is an array-wise prefix, not a string-wise-prefix:
   * e.g., `toParts(["ban"])` is not a prefix of `toParts(["banana"])`.
   */
  hasPrefix(address: Address, prefix: Address): boolean;

  /**
   * Interpret the provided string as an Address.
   *
   * Addresses are natively stored as strings. This method verifies
   * that the provided "raw" address is actually an Address, so that
   * you can have a type-level assurance that a string is an Address.
   *
   * This is useful if e.g. you are loading serialized Addresses.
   *
   * Throws an error if the string is not a valid Address.
   */
  fromRaw(raw: string): Address;

  /**
   * A parser for Addresses.
   *
   * Convenience wrapper `fromRaw`.
   */
  parser: C.Parser<Address>;
}

export type Options = {|
  /**
   * The name of this kind of address, like `NodeAddress`.
   */
  +name: string,

  /**
   * A unique nonce for the runtime representation of this address. For
   * compact serialization, this should be short; a single letter
   * suffices.
   */
  +nonce: string,

  /**
   * For the purposes of nice error messages: in response to an address
   * of the wrong kind, we can inform the user what kind of address they
   * passed (e.g., "expected NodeAddress, got EdgeAddress"). This
   * dictionary maps another address module's nonce to the name of that
   * module.
   */
  +otherNonces?: Map<string, string>,
|};

export function makeAddressModule(options: Options): AddressModule<string> {
  type Address = string; // for readability and interface consistency
  const {name, nonce} = options;
  const otherNonces = new Map(options.otherNonces || new Map());

  const separator = "\0";
  if (nonce.indexOf(separator) !== -1) {
    throw new Error(`invalid nonce (contains NUL): ${stringify(nonce)}`);
  }

  const nonceWithSeparator = nonce + separator;
  const otherNoncesWithSeparators = MapUtil.mapKeys(
    otherNonces,
    (otherNonce) => {
      if (otherNonce === nonce) {
        throw new Error(
          `primary nonce listed as otherNonce: ${stringify(nonce)}`
        );
      }
      if (otherNonce.indexOf(separator) !== -1) {
        throw new Error(
          `invalid otherNonce (contains NUL): ${stringify(otherNonce)}`
        );
      }
      return otherNonce + separator;
    }
  );

  function assertValid(address: Address, what?: string): void {
    // istanbul ignore if
    if (process.env.NODE_ENV !== "test") {
      // Catching invalid addresses in test code should be sufficient
      return;
    }
    const prefix = what == null ? "" : `${what}: `;
    if (address == null) {
      throw new Error(prefix + `expected ${name}, got: ${String(address)}`);
    }
    if (!address.endsWith(separator)) {
      throw new Error(prefix + `expected ${name}, got: ${stringify(address)}`);
    }
    if (!address.startsWith(nonceWithSeparator)) {
      for (const [
        otherNonceWithSeparator,
        otherName,
      ] of otherNoncesWithSeparators) {
        if (address.startsWith(otherNonceWithSeparator)) {
          throw new Error(
            prefix + `expected ${name}, got ${otherName}: ${stringify(address)}`
          );
        }
      }
      throw new Error(prefix + `expected ${name}, got: ${stringify(address)}`);
    }
  }

  function partsString(parts: $ReadOnlyArray<string>): string {
    // This is needed to properly print arrays containing `undefined`.
    return "[" + parts.map((p) => String(stringify(p))).join(",") + "]";
  }

  function assertValidParts(
    parts: $ReadOnlyArray<string>,
    what?: string
  ): void {
    // istanbul ignore if
    if (process.env.NODE_ENV !== "test") {
      // Catching invalid parts in test code should be sufficient
      return;
    }
    const prefix = what == null ? "" : `${what}: `;
    if (parts == null) {
      throw new Error(
        prefix + `expected array of parts, got: ${String(parts)}`
      );
    }
    parts.forEach((s: string) => {
      if (s == null) {
        throw new Error(
          prefix +
            `expected array of parts, got ${String(s)} in: ${partsString(
              parts
            )}`
        );
      }
      if (s.indexOf(separator) !== -1) {
        const where = `${stringify(s)} in ${partsString(parts)}`;
        throw new Error(prefix + `part contains NUL character: ${where}`);
      }
    });
  }

  function nullDelimited(components: $ReadOnlyArray<string>): string {
    return [...components, ""].join(separator);
  }

  function fromParts(parts: $ReadOnlyArray<string>): Address {
    assertValidParts(parts);
    return nonce + separator + nullDelimited(parts);
  }

  const empty = fromParts([]);

  function toParts(address: Address): string[] {
    assertValid(address);
    const parts = address.split(separator);
    return parts.slice(1, parts.length - 1);
  }

  function toString(address: Address): string {
    const parts = toParts(address);
    return `${name}${stringify(parts)}`;
  }

  function append(address: Address, ...parts: string[]): Address {
    assertValid(address);
    assertValidParts(parts);
    return address + nullDelimited(parts);
  }

  function hasPrefix(address: Address, prefix: Address): boolean {
    assertValid(address, "address");
    assertValid(prefix, "prefix");
    return address.startsWith(prefix);
  }

  function fromRaw(address: string): Address {
    if (!address.endsWith(separator)) {
      throw new Error(
        `address does not end with separator: ${stringify(address)}`
      );
    }
    if (!address.startsWith(nonceWithSeparator)) {
      for (const [
        otherNonceWithSeparator,
        otherName,
      ] of otherNoncesWithSeparators) {
        if (address.startsWith(otherNonceWithSeparator)) {
          throw new Error(
            `expected ${name}, got ${otherName}: ${stringify(address)}`
          );
        }
      }
      throw new Error(`expected ${name}, got: ${stringify(address)}`);
    }
    return address;
  }

  const parser: C.Parser<Address> = C.fmap(C.string, fromRaw);

  const result = {
    assertValid,
    assertValidParts,
    empty,
    fromParts,
    toParts,
    toString,
    append,
    hasPrefix,
    fromRaw,
    parser,
  };
  return deepFreeze(result);
}
