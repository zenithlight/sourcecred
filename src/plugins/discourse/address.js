// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";

import {type PostId, type TopicId, type LikeAction} from "./fetch";

import {
  topicNodeType,
  postNodeType,
  userNodeType,
  likeNodeType,
} from "./declaration";

export function topicAddress(serverUrl: string, id: TopicId): NodeAddressT {
  return NodeAddress.append(topicNodeType.prefix, serverUrl, String(id));
}

export function postAddress(serverUrl: string, id: PostId): NodeAddressT {
  return NodeAddress.append(postNodeType.prefix, serverUrl, String(id));
}

export function userAddress(serverUrl: string, username: string): NodeAddressT {
  return NodeAddress.append(userNodeType.prefix, serverUrl, username);
}

export function likeAddress(serverUrl: string, like: LikeAction): NodeAddressT {
  return NodeAddress.append(
    likeNodeType.prefix,
    serverUrl,
    like.username,
    String(like.postId)
  );
}
