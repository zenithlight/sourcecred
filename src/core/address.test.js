// @flow

import {makeAddressModule} from "./address";

describe("core/address", () => {
  describe("makeAddressModule", () => {
    const makeModules = () => ({
      FooAddress: makeAddressModule({
        name: "FooAddress",
        nonce: "F",
        otherNonces: new Map().set("B", "BarAddress"),
      }),
      BarAddress: makeAddressModule({
        name: "BarAddress",
        nonce: "B",
        otherNonces: new Map().set("F", "FooAddress"),
      }),
      WatAddress: makeAddressModule({
        name: "WatAddress",
        nonce: "W",
        otherNonces: new Map(),
      }),
    });

    it("makes an address module given the mandatory options", () => {
      makeAddressModule({name: "FooAddress", nonce: "F"});
    });
    it("makes address modules using all the options", () => {
      makeModules();
    });
    it("rejects a module whose nonce contains NUL", () => {
      expect(() => {
        makeAddressModule({name: "BadAddress", nonce: "n\0o"});
      }).toThrow("invalid nonce (contains NUL):");
    });
    it("rejects a module with `otherNonces` containing NUL", () => {
      expect(() => {
        makeAddressModule({
          name: "GoodAddress",
          nonce: "G",
          otherNonces: new Map().set("n\0o", "BadAddress"),
        });
      }).toThrow("invalid otherNonce (contains NUL):");
    });
    it("rejects a module with `nonce` in `otherNonces`", () => {
      expect(() => {
        makeAddressModule({
          name: "GoodAddress",
          nonce: "G",
          otherNonces: new Map().set("G", "WatAddress"),
        });
      }).toThrow("primary nonce listed as otherNonce");
    });
    it("returns an object with read-only properties", () => {
      const {FooAddress} = makeModules();
      expect(() => {
        // $FlowExpectedError[cannot-write]
        FooAddress.assertValid = FooAddress.assertValid; // eslint-disable-line no-self-assign
      }).toThrow(/read.only property/);
    });

    it("has a stable internal representation", () => {
      const input = ["", "hello", "", "", "sweet", "world", "", "", ""];
      const address = makeModules().FooAddress.fromParts(input);
      // We stringify the address here because otherwise literal NUL
      // characters will appear in the snapshot file.
      expect(JSON.stringify(address)).toMatchSnapshot();
    });

    describe("assertValid", () => {
      const {FooAddress, BarAddress, WatAddress} = makeModules();
      it("rejects `undefined`", () => {
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.assertValid(undefined, "widget");
        }).toThrow("widget: expected FooAddress, got: undefined");
      });
      it("rejects `null`", () => {
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.assertValid(null, "widget");
        }).toThrow("widget: expected FooAddress, got: null");
      });
      it("rejects an address from a known module", () => {
        const bar = BarAddress.fromParts(["hello"]);
        expect(() => {
          FooAddress.assertValid(bar, "widget");
        }).toThrow("widget: expected FooAddress, got BarAddress:");
      });
      it("rejects an address from an unknown module", () => {
        const wat = WatAddress.fromParts(["hello"]);
        expect(() => {
          FooAddress.assertValid(wat, "widget");
        }).toThrow("widget: expected FooAddress, got:");
      });
      it("rejects a string that starts with the nonce but no separator", () => {
        expect(() => {
          FooAddress.assertValid("F/things\0", "widget");
        }).toThrow("widget: expected FooAddress, got:");
      });
      it("rejects a string that does not end with a separator", () => {
        const address = FooAddress.fromParts(["hello"]);
        const truncated = address.substring(0, address.length - 2);
        expect(() => {
          FooAddress.assertValid(truncated, "widget");
        }).toThrow("widget: expected FooAddress, got:");
      });
      it("rejects junk", () => {
        expect(() => {
          FooAddress.assertValid("junk", "widget");
        }).toThrow("widget: expected FooAddress, got:");
      });
    });

    describe("assertValidParts", () => {
      const {FooAddress} = makeModules();
      it("rejects `undefined`", () => {
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.assertValidParts(undefined, "widget");
        }).toThrow("widget: expected array of parts, got: undefined");
      });
      it("rejects `null`", () => {
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.assertValidParts(null, "widget");
        }).toThrow("widget: expected array of parts, got: null");
      });
      it("rejects an array containing `undefined`", () => {
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.assertValidParts(["hello", undefined, "world"], "widget");
        }).toThrow(
          "widget: expected array of parts, got undefined in: " +
            '["hello",undefined,"world"]'
        );
      });
      it("rejects an array containing `null`", () => {
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.assertValidParts(["hello", null, "world"], "widget");
        }).toThrow(
          "widget: expected array of parts, got null in: " +
            '["hello",null,"world"]'
        );
      });
      it("rejects an array with a string containing a NUL character", () => {
        expect(() => {
          FooAddress.assertValidParts(["hello", "n\0o", "world"], "widget");
        }).toThrow(
          "widget: part contains NUL character: " +
            '"n\\u0000o" in ["hello","n\\u0000o","world"]'
        );
      });
    });

    describe("empty", () => {
      const {FooAddress} = makeModules();
      it("is a valid address", () => {
        FooAddress.assertValid(FooAddress.empty);
      });
      it("has empty parts", () => {
        expect(FooAddress.toParts(FooAddress.empty)).toEqual([]);
      });
    });

    describe("fromRaw", () => {
      const {FooAddress, BarAddress} = makeModules();
      function thunk(x: string) {
        return () => FooAddress.fromRaw(x);
      }
      it("throws on an empty string", () => {
        expect(thunk("")).toThrow("address does not end with separator");
      });
      it("throws on a string that doesn't start with the right separator", () => {
        expect(thunk("\0")).toThrow("expected FooAddress, got");
      });
      it("throws on a string from a different address module", () => {
        expect(thunk(BarAddress.empty)).toThrow(
          "expected FooAddress, got BarAddress"
        );
      });
      function roundTrip(x) {
        expect(FooAddress.fromRaw(x)).toEqual(x);
      }
      it("works on an empty address", () => {
        roundTrip(FooAddress.empty);
      });
      it("works on a non-empty address", () => {
        roundTrip(FooAddress.fromParts(["foo", "bar"]));
      });
    });

    describe("parser", () => {
      const {FooAddress, BarAddress} = makeModules();
      function thunk(x: string) {
        return () => FooAddress.parser.parseOrThrow(x);
      }
      it("parses a valid address", () => {
        expect(FooAddress.parser.parseOrThrow(FooAddress.empty)).toEqual(
          FooAddress.empty
        );
      });
      it("rejects an empty string", () => {
        expect(thunk("")).toThrow("address does not end with separator");
      });
      it("rejects a string that doesn't start with the right separator", () => {
        expect(thunk("\0")).toThrow("expected FooAddress, got");
      });
      it("rejects an address from the wrong module", () => {
        expect(thunk(BarAddress.empty)).toThrow(
          "expected FooAddress, got BarAddress"
        );
      });
    });

    describe("fromParts", () => {
      const {FooAddress, BarAddress} = makeModules();

      // We use this next test as a proxy for fully correct validation,
      // in conjunction with tests on `assertValid` and
      // `assertValidParts`.
      it("validates parts", () => {
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.fromParts(["hello", null, "world"]);
        }).toThrow(
          'expected array of parts, got null in: ["hello",null,"world"]'
        );
      });

      it("encodes the address kind for the empty address", () => {
        const foo = FooAddress.fromParts([]);
        const bar = BarAddress.fromParts([]);
        expect(foo).not.toEqual(bar);
      });
      it("encodes the address kind for a normal address", () => {
        const foo = FooAddress.fromParts(["hello", "world"]);
        const bar = BarAddress.fromParts(["hello", "world"]);
        expect(foo).not.toEqual(bar);
      });
    });

    describe("toParts", () => {
      const {FooAddress, BarAddress} = makeModules();

      // We use this next test as a proxy for fully correct validation,
      // in conjunction with tests on `assertValid`.
      it("validates address kind", () => {
        const bar = BarAddress.fromParts(["hello"]);
        expect(() => {
          FooAddress.toParts(bar);
        }).toThrow("expected FooAddress, got BarAddress:");
      });

      describe("is a left identity for `fromParts`", () => {
        function check(description, inputParts) {
          it(description, () => {
            const address = FooAddress.fromParts(inputParts);
            const parts = FooAddress.toParts(address);
            expect(parts).toEqual(inputParts);
          });
        }
        check("on the empty input", []);
        check("on an input made of one empty part", [""]);
        check("on an input made of lots of empty parts", ["", "", ""]);
        check("on an input with lots of empty parts throughout", [
          ...["", ""],
          "hello",
          ...["", "", ""],
          "sweet",
          "world",
          ...["", "", "", ""],
        ]);
        check("on a singleton input", ["jar"]);
        check("on an input with repeated components", ["jar", "jar"]);
      });
    });

    describe("toString", () => {
      const {FooAddress, BarAddress} = makeModules();
      const partsToString = (parts) =>
        FooAddress.toString(FooAddress.fromParts(parts));

      // We use this next test as a proxy for fully correct validation,
      // in conjunction with tests on `assertValid`.
      it("validates address kind", () => {
        const bar = BarAddress.fromParts(["hello"]);
        expect(() => {
          FooAddress.toString(bar);
        }).toThrow("expected FooAddress, got BarAddress:");
      });

      it("includes the address kind", () => {
        expect(partsToString(["hello", "world"])).toContain("FooAddress");
      });
      it("includes the address parts", () => {
        expect(partsToString(["hello", "world"])).toContain("hello");
        expect(partsToString(["hello", "world"])).toContain("world");
      });
      it("does not include any NUL characters", () => {
        expect(partsToString(["hello", "world"])).not.toContain("\0");
      });
      it("works on the empty address", () => {
        expect(partsToString([])).toEqual(expect.anything());
      });
      it("differentiates empty components", () => {
        // Each of the following should have a different `toString` form.
        const inputs = [
          partsToString([]),
          partsToString([""]),
          partsToString(["", ""]),
          partsToString(["hello", "world"]),
          partsToString(["", "hello", "world"]),
          partsToString(["", "hello", "", "world"]),
          partsToString(["", "hello", "", "", "world"]),
          partsToString(["", "hello", "", "", "world", ""]),
          partsToString(["", "hello", "", "", "world", "", ""]),
          partsToString(["", "hello", "", "", "world", "", "", ""]),
        ];
        expect(Array.from(new Set(inputs)).sort()).toHaveLength(inputs.length);
      });
    });

    describe("append", () => {
      const {FooAddress, BarAddress} = makeModules();

      // We use these next tests as a proxy for fully correct
      // validation, in conjunction with tests on `assertValid` and
      // `assertValidParts`.
      it("validates address kind", () => {
        const bar = BarAddress.fromParts(["hello"]);
        expect(() => {
          FooAddress.append(bar, "world");
        }).toThrow("expected FooAddress, got BarAddress:");
      });
      it("validates components", () => {
        const foo = FooAddress.fromParts(["hello"]);
        expect(() => {
          // $FlowExpectedError[incompatible-call]
          FooAddress.append(foo, "world", null);
        }).toThrow('expected array of parts, got null in: ["world",null]');
      });

      describe("is equivalent to the part-wise implementation on", () => {
        function check(description, baseParts, ...moreParts) {
          it(description, () => {
            const base = FooAddress.fromParts(baseParts);
            const actual = FooAddress.append(base, ...moreParts);
            const expected = FooAddress.fromParts([...baseParts, ...moreParts]);
            expect(actual).toEqual(expected);
          });
        }

        check("the null address with nothing", []);
        check("the null address with an empty component", [], "");
        check("the null address with a nonempty component", [], "a");
        check("the null address with lots of components", [], "a", "", "b");

        const base = ["a", "", "b", ""];
        check("a longer address with nothing", base);
        check("a longer address with an empty component", base, "");
        check("a longer address with a nonempty component", base, "c");
        check("a longer address with lots of components", base, "c", "", "d");
      });
    });

    describe("hasPrefix", () => {
      const {FooAddress, BarAddress} = makeModules();

      // We use these next tests as a proxy for fully correct
      // validation, in conjunction with tests on `assertValid`.
      it("validates first address kind", () => {
        const fst = BarAddress.fromParts(["hello"]);
        const snd = FooAddress.fromParts(["world"]);
        expect(() => {
          FooAddress.hasPrefix(fst, snd);
        }).toThrow("address: expected FooAddress, got BarAddress:");
      });
      it("validates second address kind", () => {
        const fst = FooAddress.fromParts(["hello"]);
        const snd = BarAddress.fromParts(["world"]);
        expect(() => {
          FooAddress.hasPrefix(fst, snd);
        }).toThrow("prefix: expected FooAddress, got BarAddress");
      });

      const {hasPrefix, fromParts} = FooAddress;
      it("accepts the empty prefix of non-empty input", () => {
        expect(hasPrefix(fromParts(["foo", "bar"]), fromParts([]))).toBe(true);
      });
      it("accepts the empty prefix of empty input", () => {
        expect(hasPrefix(fromParts([]), fromParts([]))).toBe(true);
      });
      it("rejects a non-empty prefix of empty input", () => {
        expect(hasPrefix(fromParts([]), fromParts(["foo", "bar"]))).toBe(false);
      });
      it("accepts a normal input", () => {
        expect(
          hasPrefix(fromParts(["foo", "bar", "baz"]), fromParts(["foo", "bar"]))
        ).toBe(true);
      });
      it("accepts that an address is a prefix of itself", () => {
        expect(
          hasPrefix(fromParts(["foo", "bar"]), fromParts(["foo", "bar"]))
        ).toBe(true);
      });
      it("accepts inputs with empty components", () => {
        expect(
          hasPrefix(
            fromParts(["foo", "", "bar", "", "baz"]),
            fromParts(["foo", "", "bar", ""])
          )
        ).toBe(true);
      });
      it("rejects inputs with no nontrivial common prefix", () => {
        expect(
          hasPrefix(fromParts(["foo", "bar", "baz"]), fromParts(["bar", "foo"]))
        ).toBe(false);
      });
      it("rejects inputs with insufficiently long common prefix", () => {
        expect(
          hasPrefix(
            fromParts(["foo", "bar", "baz"]),
            fromParts(["foo", "quux"])
          )
        ).toBe(false);
      });
      it("rejects when the putative prefix is a proper infix", () => {
        expect(
          hasPrefix(fromParts(["foo", "bar", "baz"]), fromParts(["bar"]))
        ).toBe(false);
      });
      it("rejects when the putative prefix is a proper suffix", () => {
        expect(
          hasPrefix(fromParts(["foo", "bar", "baz"]), fromParts(["bar", "baz"]))
        ).toBe(false);
      });
      it("rejects when the arguments are reversed", () => {
        expect(
          hasPrefix(fromParts(["foo", "bar"]), fromParts(["foo", "bar", "baz"]))
        ).toBe(false);
      });
      it("rejects when the last component is truncated", () => {
        expect(
          hasPrefix(fromParts(["foo", "bar", "baz"]), fromParts(["foo", "ba"]))
        ).toBe(false);
      });
      it("rejects when two components have been concatenated", () => {
        expect(
          hasPrefix(
            fromParts(["foo", "bar", "baz"]),
            fromParts(["foobar", "baz"])
          )
        ).toBe(false);
      });
      it("rejects an extra empty component in the middle of the base", () => {
        expect(
          hasPrefix(fromParts(["foo", "", "baz"]), fromParts(["foo", "baz"]))
        ).toBe(false);
      });
      it("rejects an extra empty component in the middle of the prefix", () => {
        expect(
          hasPrefix(fromParts(["foo", "baz"]), fromParts(["foo", "", "baz"]))
        ).toBe(false);
      });
      it("rejects an extra empty component at the end of the prefix", () => {
        expect(
          hasPrefix(fromParts(["foo", "baz"]), fromParts(["foo", "baz", ""]))
        ).toBe(false);
      });
    });
  });
});
