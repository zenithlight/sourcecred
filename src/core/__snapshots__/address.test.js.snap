// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/address makeAddressModule has a stable internal representation 1`] = `"\\"F\\\\u0000\\\\u0000hello\\\\u0000\\\\u0000\\\\u0000sweet\\\\u0000world\\\\u0000\\\\u0000\\\\u0000\\\\u0000\\""`;
